core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -769,6 +800,77 @@ private SubPlan optimizePlan(SubPlan plan)
         {
             // Re-optimize plan here based on available runtime statistics.
             // Fragments changed due to re-optimization as well as their downstream stages are expected to be assigned new fragment ids.
+            plan = updateStagesPartitioning(plan);
+            return plan;
+        }
+
+        private SubPlan updateStagesPartitioning(SubPlan plan)
+        {
+            if (!runtimeAdaptivePartitioningEnabled || runtimeAdaptivePartitioningApplied) {
+                return plan;
+            }
+
+            for (SubPlan subPlan : planInTopologicalOrder) {
+                PlanFragment fragment = subPlan.getFragment();
+                if (!hasInputPartitioning(fragment)) {
+                    // no input hash partitioning present
+                    continue;
+                }
+
+                StageId stageId = getStageId(fragment.getId());
+                if (stageExecutions.containsKey(stageId)) {
+                    // already started
+                    continue;
+                }
+
+                IsReadyForExecutionResult isReadyForExecutionResult = isReadyForExecution(subPlan);
+                // caching is not only needed to avoid duplicate calls, but also to avoid the case
+                // that a subPlan that is not ready now but becomes ready when updateStageExecutions
```

Comment:
Could you please elaborate on this?

Reply:
Discussed offline:

We want to make sure partition count is updated before stage execution is started. We want to avoid starting an execution without considering changing the number of partitions.